presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/DatabaseMetadataWriter.java

```diff
@@ -0,0 +1,728 @@
+    }
+
+    private static void insertTable(MasterWriterDao dao, long commitId, TableInfo table)
+    {
+        dao.insertTable(
+                commitId,
+                table.getTableId(),
+                table.getTableName().getBytes(UTF_8),
+                table.getSchemaId(),
+                table.getDistributionId(),
+                boxedLong(table.getTemporalColumnId()),
+                table.getCreateTime(),
+                table.getUpdateTime(),
+                table.getRowCount(),
+                utf8Bytes(table.getComment()));
+    }
+
+    private static void insertColumn(MasterWriterDao dao, long commitId, long tableId, ColumnInfo column)
+    {
+        dao.insertColumn(
+                commitId,
+                column.getColumnId(),
+                column.getColumnName().getBytes(UTF_8),
+                column.getType().getTypeSignature().toString().getBytes(UTF_8),
+                tableId,
+                column.getOrdinal(),
+                boxedInt(column.getBucketOrdinal()),
+                boxedInt(column.getSortOrdinal()),
+                utf8Bytes(column.getComment()));
+    }
+
+    private static void updateTable(MasterWriterDao dao, long commitId, long tableId)
+    {
+        TableInfo table = dao.getTableInfo(tableId).builder()
+                .setUpdateTime(System.currentTimeMillis())
+                .build();
+
+        deleteTable(dao, commitId, tableId);
+        insertTable(dao, commitId, table);
+    }
+
+    private static void deleteSchema(MasterWriterDao dao, long commitId, long schemaId)
+    {
+        int rows = dao.deleteSchema(commitId, schemaId);
+        verifyMetadata(rows == 1, "Wrong row count %s for delete of schema (%s)", rows, schemaId);
+    }
+
+    private static void deleteTable(MasterWriterDao dao, long commitId, long tableId)
+    {
+        int rows = dao.deleteTable(commitId, tableId);
+        verifyMetadata(rows == 1, "Wrong row count %s for delete of table (%s)", rows, tableId);
+    }
+
+    private static void deleteView(MasterWriterDao dao, long commitId, long viewId)
+    {
+        int rows = dao.deleteView(commitId, viewId);
+        verifyMetadata(rows == 1, "Wrong row count %s for delete of view (%s)", rows, viewId);
+    }
+
+    private static void deleteColumn(MasterWriterDao dao, long commitId, long tableId, long columnId)
+    {
+        int rows = dao.deleteColumn(commitId, tableId, columnId);
+        verifyMetadata(rows == 1, "Wrong row count %s for delete of column (%s:%s)", rows, tableId, columnId);
+    }
+
+    private static void deleteChunks(ShardWriterDao dao, long commitId, long tableId, Set<Long> chunkIds)
+    {
+        int rows = dao.deleteChunks(commitId, chunkIds);
+        verifyMetadata(rows == chunkIds.size(), "Wrong row count %s for delete of %s chunks", rows, chunkIds.size());
+
+        rows = dao.deleteIndexChunks(chunkIndexTable(tableId), chunkIds);
```

Comment:
Here we delete all, instead of setting end_commit_id. The following 2 cases may be issues:
1  ChunkSupplier.java line 105, getBucketChunks.   called by RaptorSplitManager.getSplits(), the only place  for presto to get chunks. If it can be called twice in a transaction, not good. But if only called once, maybe fine.

2 DatabaseMetadataWriter.recover line 205 dao.rollbackDeletedIndexChunks(chunkIndexTable(tableId), commitId);  filters end_commit_id. So the rollback of delete,  only rollback chunks, but index are not, then index is less then chunks.

Reply:
Good catch, this is a bug.